File: rhododendron/lib/rhododendron/utils/ssha512.ex
defmodule Rhododendron.Ssha512 do
  # https://mad9scientist.com/dovecot-password-creation-php/
  # https://doc.dovecot.org/2.3/configuration_manual/howto/convert_password_schemes/
  def sign(data, salt_len) do
    salt = :crypto.strong_rand_bytes(salt_len)
    header() <> ((:crypto.hash(:sha512, data <> salt) <> salt) |> Base.encode64())
  end

  def verify(hashed, data) do
    if(String.starts_with?(hashed, header())) do
      {:ok, <<tmp::binary-size(64), salt::binary>>} =
        String.trim_leading(hashed, header()) |> Base.decode64()

      if :crypto.hash(:sha512, data <> salt) == tmp do
        {:ok}
      else
        {:error, "invalid salt"}
      end
    else
      {:error, "invalid ssha-512 message"}
    end
  end

  defp header do
    "{SSHA512}"
  end
end

File: portal/src/ssha512.rs
use data_encoding::BASE64;
use openssl::{hash::MessageDigest, sha::Sha512};

use super::random::bytes as random_bytes;

// https://mad9scientist.com/dovecot-password-creation-php/
// https://doc.dovecot.org/2.3/configuration_manual/howto/convert_password_schemes/
pub fn sign(password: &str, salt_len: usize) -> String {
    let salt = random_bytes(salt_len);
    let mut buf = Vec::new();
    let hash = {
        let mut hashed = Sha512::new();
        hashed.update(password.as_bytes());
        hashed.update(&salt);
        hashed.finish()
    };
    buf.extend_from_slice(&hash);
    buf.extend_from_slice(&salt);
    format!("{}{}", PREFIX, BASE64.encode(&buf))
}
pub fn verify(code: &str, password: &str) -> bool {
    let hash_size = {
        let it = MessageDigest::sha512();
        it.size()
    };
    if let Some(code) = code.strip_prefix(PREFIX)
        && let Ok(ref buf) = BASE64.decode(code.as_bytes())
        && buf.len() > hash_size
    {
        let tmp = {
            let salt = &buf[hash_size..];
            let mut hashed = Sha512::new();
            hashed.update(password.as_bytes());
            hashed.update(salt);
            hashed.finish()
        };
        return buf[0..hash_size] == tmp;
    }

    false
}

static PREFIX: &str = "{SSHA512}";
